fix(common): incorrect error type for XHR errors in `TestRequest`

devversion opened the pull request:
Currently the `HttpClient` always wraps errors from XHR requests, but
the underlying errors are always of type `ProgressEvent`, or don't have
a native error if the status code is just indicating failure (e.g. 404).

This behavior does not match in the `TestRequest` class provided by
`@angular/common/http/testing` where errors are considered being
of type `ErrorEvent`. This is incorrect because `ErrorEvent`s provide
information for errors in scripts or files which are evaluated. Since
the `HttpClient` never evaluates scripts/files, and also since XHR requests
clearly are documented to emit `ProgressEvent`'s, we should change the
`TestSupport` to retrieve such `ProgressEvent`'s instead of incompatible
objects of type `ErrorEvent`.

Resources:
  * https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/error_event
  * https://developer.mozilla.org/en-US/docs/Web/API/ErrorEvent
  * https://xhr.spec.whatwg.org/#event-xhr-errpr

Related to: https://github.com/angular/angular/issues/34748.

jelbourn reviewed (approved):
LGTM - does anything need to go into the deprecation doc?

Reviewed-for: public-api

mary-poppins commented:
You can preview 240ed96 at https://pr36082-240ed96.ngbuilds.io/.
You can preview e7d54f4 at https://pr36082-e7d54f4.ngbuilds.io/.
You can preview db6d39f at https://pr36082-db6d39f.ngbuilds.io/.
You can preview 7ecec02 at https://pr36082-7ecec02.ngbuilds.io/.
You can preview cd1d50f at https://pr36082-cd1d50f.ngbuilds.io/.
You can preview 6c52c2f at https://pr36082-6c52c2f.ngbuilds.io/.
You can preview 5acd0f2 at https://pr36082-5acd0f2.ngbuilds.io/.
You can preview 58698c6 at https://pr36082-58698c6.ngbuilds.io/.

devversion commented:
@jelbourn Yeah, good point. Added an entry to the deprecation guide.

mary-poppins commented:
You can preview 3cac51b at https://pr36082-3cac51b.ngbuilds.io/.

mary-poppins commented:
You can preview e6a238c at https://pr36082-e6a238c.ngbuilds.io/.

devversion committed:
fix(common): incorrect error type for XHR errors in `TestRequest`

Currently the `HttpClient` always wraps errors from XHR requests, but
the underlying errors are always of type `ProgressEvent`, or don't have
a native error if the status code is just indicating failure (e.g. 404).

This behavior does not match in the `TestRequest` class provided by
`@angular/common/http/testing` where errors are considered being
of type `ErrorEvent`. This is incorrect because `ErrorEvent`s provide
information for errors in scripts or files which are evaluated. Since
the `HttpClient` never evaluates scripts/files, and also since XHR requests
clearly are documented to emit `ProgressEvent`'s, we should change the
`TestSupport` to retrieve such `ProgressEvent`'s instead of incompatible
objects of type `ErrorEvent`.

In favor of having a deprecation period, we keep supporting `ErrorEvent`
in the `TestRequest.error` signature. Eventually, we can remove this
signature in the future.

Resources:
  * https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/error_event
  * https://developer.mozilla.org/en-US/docs/Web/API/ErrorEvent
  * https://xhr.spec.whatwg.org/#event-xhr-errpr

Related to: https://github.com/angular/angular/issues/34748.

DEPRECATED: `TestRequest` from `@angular/common/http/testing` no longer
accepts `ErrorEvent` when simulating XHR errors. Instead instances of
`ProgressEvent` should be passed, matching with the native browser behavior.

mary-poppins commented:
You can preview 40335db at https://pr36082-40335db.ngbuilds.io/.

jelbourn commented:
Adding @aikidave for the deprecation guide

mary-poppins commented:
You can preview 618e643 at https://pr36082-618e643.ngbuilds.io/.

devversion commented:
FYI: CI is red due to failures in `master`: https://github.com/angular/angular/pull/37679, but PR is still ready for review. I'll rebase when there is more feedback / or it's ready for merge.

jelbourn commented on aio/content/guide/deprecations.md:
```diff
@@ -489,6 +490,30 @@ If you rely on the behavior that the same object instance should cause change de
+{@a testrequest-errorevent}
+### `TestRequest` accepting `ErrorEvent`
+
+The `@angular/common` package provides utilities for testing the `HttpClient`. One part of these
+testing utilities is the `TestRequest` class. It serves as mock for HTTP requests and is primarily
+used in the `HttpTestingController`.
```
```suggestion
The `@angular/common` package provides utilities for testing `HttpClient`. The `TestRequest` class from `@angular/common/http/testing` mocks HTTP request objects for use with `HttpTestingController.
```

devversion replied:
Yeah, that's definitely better. thx!

jelbourn commented on aio/content/guide/deprecations.md:
```diff
@@ -489,6 +490,30 @@ If you rely on the behavior that the same object instance should cause change de
+testing utilities is the `TestRequest` class. It serves as mock for HTTP requests and is primarily
+used in the `HttpTestingController`.
+
+`TestRequest` provides an interface for resolving to mocked content, or simulating an errored http
+request. Currently, the method for simulating an error accepts objects of type `ErrorEvent`. This
+does not match with error events that are returned by browsers natively, so we intend to no longer
+accept `ErrorEvent`.
```
```suggestion
`TestRequest` provides an API for simulating an HTTP response with an error. In earlier versions of Angular, this API accepted objects of type `ErrorEvent`, which
does not match the type of error event that are returned by browsers natively. If you use `ErrorEvent` with `TestRequest`, you should switch to `ProgressEvent`. A future version of Angular will remove support for using `ErrorEvent` with `TestRequest`. 
```

devversion replied:
Thanks! way better 👍

alfaproject commented on goldens/public-api/common/http/testing/testing.d.ts:
```diff
@@ -25,13 +25,8 @@ export declare class TestRequest {
-        statusText?: string;
-    }): void;
+    /** @deprecated */ error(error: ErrorEvent, opts?: TestRequestErrorOptions): void;
+    error(error: ProgressEvent, opts?: TestRequestErrorOptions): void;
```
`ProgressEvent` sounds very specific. Is this intended?

devversion replied:
@alfaproject yeah, that is true. What would you expect in those cases? The goal here was to match with native error being constructed when a request fails. https://xhr.spec.whatwg.org/#event-xhr-error

It seems ambiguous that `TestRequest.error` could theoretically also deal with non-network level errors. e.g. just a simple 404 error. In that case though it's unclear what `ErrorEvent` or `ProgressEvent` (with this change) should be though. Ideally only network errors are handled here and other server responses are simulated with `flush`. This is ambiguous though as the method logic supports status responses, status text and headers.

alfaproject replied:
I guess I was expecting to see the `Event` base class in the public API but I suppose that's too broad as well. A `ProgressEvent` is not necessarily an error and that's why I got a bit confused.

`ErrorEvent` also extends from `Event` but as you mentioned that's not really emitted.

¯\_(ツ)_/¯

mary-poppins commented:
You can preview 3269acb at https://pr36082-3269acb.ngbuilds.io/.

jelbourn commented on aio/content/guide/deprecations.md:
```diff
@@ -489,6 +490,29 @@ If you rely on the behavior that the same object instance should cause change de
+you should switch to `ProgressEvent`. A future version of Angular will remove support for
+using `ErrorEvent` with `TestRequest`.
+
+Here is an example using a `ProgressEvent`:
```
```suggestion
The following example demonstrates use of `ProgressEvent` with `HttpTestingController`:
```

jelbourn reviewed (approved):
LGTM

Reviewed-for: public-api

IgorMinar reviewed (approved):
the deprecation lgtm

Reviewed-for: docs-packaging-and-releasing

IgorMinar reviewed (approved):
the deprecation lgtm

Reviewed-for: docs-packaging-and-releasing, public-api

pkozlowski-opensource reviewed (commented):
LGTM overall but I think that @alxhub has more context so I'm not going to submit this as approving review (so Alex has a chance of rubber-stamping it).

pkozlowski-opensource reviewed (approved):
Oh, actually @alxhub reviewed this already. Again, this LGTM.

Reviewed-for: public-api

aikidave commented on aio/content/guide/deprecations.md:
```diff
@@ -489,6 +490,29 @@ If you rely on the behavior that the same object instance should cause change de
+
+`TestRequest` provides an API for simulating an HTTP response with an error. In earlier versions
+of Angular, this API accepted objects of type `ErrorEvent`, which does not match the type of
+error event being returned by browsers natively. If you use `ErrorEvent` with `TestRequest`,
```
```suggestion
error event that browsers return natively. If you use `ErrorEvent` with `TestRequest`,
```
Passive voice. Hard to avoid sometimes!

devversion replied:
thanks. that's much better 👍

aikidave commented on aio/content/guide/deprecations.md:
```diff
@@ -489,6 +490,29 @@ If you rely on the behavior that the same object instance should cause change de
+`TestRequest` provides an API for simulating an HTTP response with an error. In earlier versions
+of Angular, this API accepted objects of type `ErrorEvent`, which does not match the type of
+error event being returned by browsers natively. If you use `ErrorEvent` with `TestRequest`,
+you should switch to `ProgressEvent`. A future version of Angular will remove support for
+using `ErrorEvent` with `TestRequest`.
```
```suggestion
you should switch to `ProgressEvent`.
```

In general, I advise against statements such as "a future version..." or "in the future..." They often don't add as much value as we think--often, readers care about the state of the product as it is right now, not how it will be at some point in the future.

In this case, I think there's a second reason to remove this sentence: We're already in the deprecations topic. The fact that something might be removed is probably well understood. :)

aikithoughts reviewed (approved):
Just a few suggestions. Otherwise, LGTM.

Reviewed-for: global-docs-approvers

jessicajaniuk commented:
@devversion This PR was approved, but looks like it stalled. Can you take a look at it again?

devversion commented on aio/content/examples/http/src/testing/http-client.spec.ts:
```diff
@@ -138,30 +138,19 @@ describe('HttpClient testing', () => {
       data => fail('should have failed with the network error'),
       (error: HttpErrorResponse) => {
-        expect(error.error.message).toEqual(emsg, 'message');
+        expect(error.error).toBe(mockError);
```
I think this is not necessarily related to this PR, but I added the `done` completion. seems like a good improvement.

mary-poppins commented:
You can preview 3e29972 at https://pr36082-3e29972.ngbuilds.io/.

devversion commented:
@jessicajaniuk I've just resurrected this PR. Addressed feedback etc. I just need one more public API approval.

jessicajaniuk reviewed (approved):
LGTM :cookie:

reviewed-for: public-api

jessicajaniuk commented:
I'll run a TGP to make sure we're safe in google3 with this change.

jessicajaniuk commented:
@devversion Can you rebase this and repush? It seems to fail to build in google3 now.

mary-poppins commented:
You can preview 9598e7c at https://pr36082-9598e7c.ngbuilds.io/.

jessicajaniuk commented:
The TGP passed.

jessicajaniuk commented:
This PR was merged into the repository by commit 489cf42cd07ac2b549ea958127590b30f1b3d28f.
